src/collector/tweak_score_top_collector.rs

```diff
@@ -76,49 +128,229 @@ where
     }
 }
 
-pub struct TopTweakedScoreSegmentCollector<TSegmentScoreTweaker, TScore>
-where
-    TScore: 'static + PartialOrd + Clone + Send + Sync + Sized,
-    TSegmentScoreTweaker: ScoreSegmentTweaker<TScore>,
+pub struct TopTweakedScoreSegmentCollector<TSegmentScoreTweaker>
+where TSegmentScoreTweaker: ScoreSegmentTweaker
 {
-    segment_collector: TopSegmentCollector<TScore>,
+    segment_collector: TopSegmentCollector<TSegmentScoreTweaker::SegmentScore>,
     segment_scorer: TSegmentScoreTweaker,
 }
 
-impl<TSegmentScoreTweaker, TScore> SegmentCollector
-    for TopTweakedScoreSegmentCollector<TSegmentScoreTweaker, TScore>
-where
-    TScore: 'static + PartialOrd + Clone + Send + Sync,
-    TSegmentScoreTweaker: 'static + ScoreSegmentTweaker<TScore>,
+impl<TSegmentScoreTweaker> SegmentCollector
+    for TopTweakedScoreSegmentCollector<TSegmentScoreTweaker>
+where TSegmentScoreTweaker: 'static + ScoreSegmentTweaker
 {
-    type Fruit = Vec<(TScore, DocAddress)>;
+    type Fruit = Vec<(TSegmentScoreTweaker::Score, DocAddress)>;
 
     fn collect(&mut self, doc: DocId, score: Score) {
-        let score = self.segment_scorer.score(doc, score);
-        self.segment_collector.collect(doc, score);
+        // Thanks to generics, this if-statement is free.
+        if TSegmentScoreTweaker::is_lazy() {
+            self.segment_collector
+                .topn_computer
+                .push_lazy(doc, score, &mut self.segment_scorer);
```

Comment:
This assumes that the body of `TopSegmentCollector::collect` is empty other than this call, which is a bit fragile.

Does that suggest that `Collector::collect` should have a lazy variant too? I can't imagine what that would look like.